client.go

```diff
@@ -2,190 +2,159 @@ package documentdb
-	Create(link string, body, ret interface{}) error
-	Upsert(link string, body, ret interface{}) error
-	Replace(link string, body, ret interface{}) error
-	Execute(link string, body, ret interface{}) error
+	Read(link string, ret interface{}, opts ...CallOption) (*Response, error)
+	Delete(link string, opts ...CallOption) (*Response, error)
+	Query(link string, query *Query, ret interface{}, opts ...CallOption) (*Response, error)
+	Create(link string, body, ret interface{}, opts ...CallOption) (*Response, error)
+	Upsert(link string, body, ret interface{}, opts ...CallOption) (*Response, error)
+	Replace(link string, body, ret interface{}, opts ...CallOption) (*Response, error)
+	Execute(link string, body, ret interface{}, opts ...CallOption) (*Response, error)
 }
 
 type Client struct {
 	Url    string
-	Config Config
+	Config *Config
 	http.Client
 }
 
-// Read resource by self link
-func (c *Client) Read(link string, ret interface{}) error {
-	return c.method("GET", link, http.StatusOK, ret, &bytes.Buffer{})
-}
-
-// Read resource by self link and with request options
-func (c *Client) ReadWithRequestOptions(link string, ret interface{}, requestOptions []func(*RequestOptions)) error {
-	req, err := http.NewRequest("GET", path(c.Url, link), &bytes.Buffer{})
-	if err != nil {
-		return err
-	}
-	r := ResourceRequest(link, req)
+func (c *Client) apply(r *Request, opts []CallOption) (err error) {
 	if err = r.DefaultHeaders(c.Config.MasterKey); err != nil {
 		return err
 	}
-	if err = r.RequestOptionsHeaders(requestOptions); err != nil {
-		return err
+
+	for i := 0; i < len(opts); i++ {
+		if err = opts[i](r); err != nil {
+			return err
+		}
 	}
-	return c.do(r, http.StatusOK, ret)
+	return nil
+}
+
+// Read resource by self link
+func (c *Client) Read(link string, ret interface{}, opts ...CallOption) (*Response, error) {
+	buf := buffers.Get().(*bytes.Buffer)
+	buf.Reset()
+	res, err := c.method("GET", link, http.StatusOK, ret, buf, opts...)
+
+	buffers.Put(buf)
+
+	return res, err
 }
 
 // Delete resource by self link
-func (c *Client) Delete(link string) error {
-	return c.method("DELETE", link, http.StatusNoContent, nil, &bytes.Buffer{})
+func (c *Client) Delete(link string, opts ...CallOption) (*Response, error) {
+	return c.method("DELETE", link, http.StatusNoContent, nil, &bytes.Buffer{}, opts...)
 }
 
 // Query resource
-func (c *Client) Query(link, query string, ret interface{}) error {
-	buf := bytes.NewBufferString(querify(query))
-	req, err := http.NewRequest("POST", path(c.Url, link), buf)
-	if err != nil {
-		return err
-	}
-	r := ResourceRequest(link, req)
-	if err = r.DefaultHeaders(c.Config.MasterKey); err != nil {
-		return err
+func (c *Client) Query(link string, query *Query, ret interface{}, opts ...CallOption) (*Response, error) {
+	var (
+		err error
+		req *http.Request
+		buf = buffers.Get().(*bytes.Buffer)
+	)
+	buf.Reset()
+	defer func() { buffers.Put(buf) }()
```

Comment:
Why not like this? Do I miss something?
```suggestion
	defer buffers.Put(buf)
```